e2e/cluster_proxy_test.go

```diff
@@ -0,0 +1,272 @@
+	donec    chan struct{}
+
+	proc *expect.ExpectProcess
+}
+
+func (pp *proxyProc) endpoints() []string { return []string{pp.ep} }
+
+func (pp *proxyProc) start() error {
+	if pp.proc != nil {
+		panic("already started")
+	}
+	proc, err := spawnCmd(append([]string{pp.execPath}, pp.args...))
+	if err != nil {
+		return err
+	}
+	pp.proc = proc
+	return nil
+}
+
+func (pp *proxyProc) waitReady(readyStr string) error {
+	defer close(pp.donec)
+	return waitReadyExpectProc(pp.proc, []string{readyStr})
+}
+
+func (pp *proxyProc) Stop() error {
+	if pp.proc == nil {
+		return nil
+	}
+	if err := pp.proc.Stop(); err != nil && !strings.Contains(err.Error(), "exit status 1") {
+		// v2proxy exits with status 1 on auto tls; not sure why
+		return err
+	}
+	pp.proc = nil
+	<-pp.donec
+	pp.donec = make(chan struct{})
+	return nil
+}
+
+func (pp *proxyProc) WithStopSignal(sig os.Signal) os.Signal {
+	ret := pp.proc.StopSignal
+	pp.proc.StopSignal = sig
+	return ret
+}
+
+func (pp *proxyProc) Close() error { return pp.Stop() }
+
+type proxyV2Proc struct {
+	proxyProc
+	dataDir string
+}
+
+func proxyListenURL(cfg *etcdServerProcessConfig, portOffset int) string {
+	u, err := url.Parse(cfg.acurl)
+	if err != nil {
+		panic(err)
+	}
+	host, port, _ := net.SplitHostPort(u.Host)
+	p, _ := strconv.ParseInt(port, 10, 16)
+	u.Host = fmt.Sprintf("%s:%d", host, int(p)+portOffset)
+	return u.String()
+}
+
+func newProxyV2Proc(cfg *etcdServerProcessConfig) *proxyV2Proc {
+	listenAddr := proxyListenURL(cfg, 2)
+	name := fmt.Sprintf("testname-proxy-%p", cfg)
+	args := []string{
+		"--name", name,
+		"--proxy", "on",
+		"--listen-client-urls", listenAddr,
+		"--initial-cluster", cfg.name + "=" + cfg.purl.String(),
+	}
+	return &proxyV2Proc{
+		proxyProc{
+			execPath: cfg.execPath,
+			args:     append(args, cfg.tlsArgs...),
+			ep:       listenAddr,
+			donec:    make(chan struct{}),
+		},
+		name + ".etcd",
+	}
+}
+
+func (v2p *proxyV2Proc) Start() error {
+	os.RemoveAll(v2p.dataDir)
+	if err := v2p.start(); err != nil {
+		return err
+	}
+	return v2p.waitReady("httpproxy: endpoints found")
+}
+
+func (v2p *proxyV2Proc) Restart() error {
+	if err := v2p.Stop(); err != nil {
+		return err
+	}
+	return v2p.Start()
+}
+
+func (v2p *proxyV2Proc) Stop() error {
+	if err := v2p.proxyProc.Stop(); err != nil {
+		return err
+	}
+	// v2 proxy caches members; avoid reuse of directory
+	return os.RemoveAll(v2p.dataDir)
+}
+
+type proxyV3Proc struct {
+	proxyProc
+}
+
+func newProxyV3Proc(cfg *etcdServerProcessConfig) *proxyV3Proc {
+	listenAddr := proxyListenURL(cfg, 3)
+	args := []string{
+		"grpc-proxy",
+		"start",
+		"--listen-addr", strings.Split(listenAddr, "/")[2],
+		"--endpoints", cfg.acurl,
+		// pass-through member RPCs
+		"--advertise-client-url", "",
+	}
+	tlsArgs := []string{}
+	for i := 0; i < len(cfg.tlsArgs); i++ {
+		switch cfg.tlsArgs[i] {
+		case "--cert-file":
+			tlsArgs = append(tlsArgs, "--cert", cfg.tlsArgs[i+1], "--cert-file", cfg.tlsArgs[i+1])
```

Comment:
--cert is proxy <-> etcd (in grpcproxy already), --cert-file is client <-> proxy (new)